text/0000-pmgr.md

```diff
@@ -0,0 +1,257 @@
+temporarily map pages into accessable memory.
+
+### Request Types
+
+When communicating lists of pages back and forth, both sides use an array of `PhysRange`.
+The `PhysRange` struct looks like: 
+
+```{rust}
+struct PhysRange {
+    start: u64,
+    len: u32,
+    _flags: u32,
+}
+```
+
+Thus each side can communicate sets of contiguous pages with each other. The flags field is currently unused, but when
+used should be replaced with a newtype that has the same representation as a u32.
+
+## Object Info Requests
+
+The object info struct looks like:
+
+```{rust}
+struct ObjectInfo {
+    lifetime: LifetimeType,
+    backing: BackingType,
+}
+```
+
+This struct will be expanded over-time.
+
+## Eviction
+
+To support eviction, the kernel can inform the pager that a page must by written back to stable
+storage. The pager must honor the kernel's command and must assume the page is dirty. However, this
+is meant to be used as a last resort. The kernel can also submit a *request* that the pager free up
+some number of pages. If a page returned to the kernel is dirty, the kernel will submit an
+eviction notice for that page to inform the pager that it needs to write-back that page.
+
+Note that there is a distinction between "Evict" and "Sync". The former alone requires that data be stored but not
+overwrite object data that may have been modified in-memory. This is required to support transactional semantics well.
+We can combine Sync and Evict to recover standard OS eviction semantics. Sync by itself just syncs and doesn't
+evict. Finally, the pager can request the kernel evict object memory.
+
+## Object Page Requests
+
+The kernel can request the pager to fill a DRAM page with data from stable storage and return it to the kernel. The 
+`ObjectRange` struct looks like: 
+
+```{rust}
+struct ObjectRange {
+    start: u64,
+    len: u32,
+    _flags: u32,
+}
+```
+
+Thus each side can communicate sets of contiguous byte-granularity regions of an object with each other. The flags field is currently unused, but when
+used should be replaced with a newtype that has the same representation as a u32. Note that this means syncing and eviction can
+happen at byte-level granularity, however it's an error to submit a page request that is non-page-aligned. What does
+byte-level granularity eviction mean? It means that region needs to reset to stable-storage, including bytes that may
+not cover a full page.
+
+When the kernel requests information about a page, it sends a set of `ObjectRange` structs for a given object ID. The
+pager responds with a set of `PhysRange` structs. The pager's response does not need to fulfill the entire request, as
+long as at least one page is filled (probably the first one). The kernel reads out physical pages, in order as specified
+by the completion, and maps them in sequence into the object's page tree.
+
+The pager, at any time, can prefetch (if it has memory) and inform the kernel of its prefetching work. It does this via
```

Comment:
[minor] If prefetching is not associated with a concrete `KernelRequest` and is just a heads-up operation, does it make sense for it to be submitted over the `PagerRequest` queue?

Reply:
I think that's what I intended, yeah